metrics.go

```diff
@@ -0,0 +1,131 @@
+// limitations under the License.
+
+package stackdriver
+
+/*
+The code in this file is responsible for converting OpenCensus Proto metrics
+directly to Stackdriver Metrics.
+*/
+
+import (
+	"fmt"
+
+	"github.com/golang/protobuf/ptypes/timestamp"
+
+	distributionpb "google.golang.org/genproto/googleapis/api/distribution"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+func fromProtoPoint(startTime *timestamp.Timestamp, pt *metricspb.Point) (*monitoringpb.Point, error) {
+	if pt == nil {
+		return nil, nil
+	}
+
+	mptv, err := protoToMetricPoint(pt.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	mpt := &monitoringpb.Point{
+		Value: mptv,
+		Interval: &monitoringpb.TimeInterval{
+			StartTime: startTime,
+			EndTime:   pt.Timestamp,
+		},
+	}
+	return mpt, nil
+}
+
+func protoToMetricPoint(value interface{}) (*monitoringpb.TypedValue, error) {
+	if value == nil {
+		return nil, nil
+	}
+
+	var err error
+	var tval *monitoringpb.TypedValue
+	switch v := value.(type) {
+	default:
+		// All the other types are not yet handled.
+		// TODO: (@odeke-em, @songy23) talk to the Stackdriver team to determine
+		// the use cases for:
+		//
+		//      *TypedValue_BoolValue
+		//      *TypedValue_StringValue
+		//
+		// and then file feature requests on OpenCensus-Specs and then OpenCensus-Proto,
+		// lest we shall error here.
+		err = fmt.Errorf("protoToMetricPoint: unknown Data type: %T", value)
+
+	case *metricspb.Point_Int64Value:
+		tval = &monitoringpb.TypedValue{
+			Value: &monitoringpb.TypedValue_Int64Value{
+				Int64Value: v.Int64Value,
+			},
+		}
+
+	case *metricspb.Point_DoubleValue:
+		tval = &monitoringpb.TypedValue{
+			Value: &monitoringpb.TypedValue_DoubleValue{
+				DoubleValue: v.DoubleValue,
+			},
+		}
+
+	case *metricspb.Point_DistributionValue:
+		dv := v.DistributionValue
+		var mv *monitoringpb.TypedValue_DistributionValue
+		if dv != nil {
+			var mean float64
+			if dv.Count > 0 {
+				mean = float64(dv.Sum) / float64(dv.Count)
+			}
+			mv = &monitoringpb.TypedValue_DistributionValue{
+				DistributionValue: &distributionpb.Distribution{
+					Count:                 dv.Count,
+					Mean:                  mean,
+					SumOfSquaredDeviation: dv.SumOfSquaredDeviation,
+					BucketCounts:          bucketCounts(dv.Buckets),
+				},
+			}
+
+			if bopts := dv.BucketOptions; bopts != nil && bopts.Type != nil {
+				bexp, ok := bopts.Type.(*metricspb.DistributionValue_BucketOptions_Explicit_)
+				if ok && bexp != nil && bexp.Explicit != nil {
+					mv.DistributionValue.BucketOptions = &distributionpb.Distribution_BucketOptions{
+						Options: &distributionpb.Distribution_BucketOptions_ExplicitBuckets{
+							ExplicitBuckets: &distributionpb.Distribution_BucketOptions_Explicit{
+								Bounds: bexp.Explicit.Bounds[:],
+							},
+						},
+					}
+				}
+			}
+		}
+		tval = &monitoringpb.TypedValue{Value: mv}
+	}
+
```

Comment:
Thanks for raising this @songy23! When converting from `view.Data` to `metricspb.Metric`, I translated `view.SumData` this into one of the `metricspb.Point_Int64Value` and `metricspb.Point_DoubleValue` but this now lets me know that perhaps I made a mistake and that I should instead translate `view.SumData` to `metricspb.Point_SummaryValue`

Reply:
There's a [`Summary` value type](https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/metrics/v1/metrics.proto#L268) in our metrics definition. Although the OC `Stats` library doesn't produce `Summary` values, it could be from other sources (e.g Dropwizard) and be exported via OC Stackdriver exporter, and we should handle that. See the corresponding PR in Java: https://github.com/census-instrumentation/opencensus-java/pull/1591.

Reply:
Gotcha, I was actually working on the SummaryValue extraction here so sure I'll make that update, thanks @songy23!

Reply:
I've updated with support for SummaryValue, PTAL @songy23!

Reply:
We used special label keys and suffixes for `Summary` values: https://github.com/census-instrumentation/opencensus-java/pull/1591/files#diff-d2886f5b11a5e3be7a00a8f5fff5a1f0R89. You can leave a TODO here and update in a following PR, since that's a lot of code changes.

Reply:
Can we perhaps handle that in a separate PR, seems a bit complex and currently even out of scope as the OpenCensus-Go Stackdriver exporter doesn't support that conversion currently. I'll just add a TODO and file an issue.

Reply:
SGTM. Please also revert https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/pull/65/commits/3263075a0664985878bef067a02ee7a236899ea6 for now.